File: python/src/core/cert.py
from os import path
import requests

from config import SERVER_HOST

__all__ = ['create_certificate', 'init_certificate']


CERT_ENDPOINT: str = f'https://{SERVER_HOST}/cert'


def parent_dir(filename):
    """
    Fetch the filename from the parent directory of this file.
    :param filename: The filename to fetch
    """

    return path.join(path.dirname(__file__), '..', filename)


def create_certificate():
    """
    Create a certificate by requesting a new one from the server and writing to the root directory.
    """

    response = requests.post(CERT_ENDPOINT, verify=False)

    if response.status_code == 200:
        data = response.json()

        private_key = data['private']
        public_key = data['public']
        certificate = data['cert']

        with open(parent_dir('player.key'), 'w', encoding='utf-8') as f:
            f.write(private_key)
        with open(parent_dir('player.pub'), 'w', encoding='utf-8') as f:
            f.write(public_key)
        with open(parent_dir('player.crt'), 'w', encoding='utf-8') as f:
            f.write(certificate)


def init_certificate():
    """
    Initialize the player certificate by checking if it exists. If it does not,
    a new certificate will be fetched from the server before starting.
    """

    key = parent_dir('player.key')
    cert = parent_dir('player.crt')

    if not (path.exists(key) and path.exists(cert)):
        import urllib3

        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

        create_certificate()

        print('Created player.key, player.pub and player.crt certificate files')

File: javascript/src/core/cert.js
const fetch = require('node-fetch')
const fs = require('fs')
const path = require('path')

const config = require('../config')


const CERT_ENDPOINT = `https://${config.SERVER_HOST}/cert`


/**
 * Fetch the filename from the parent directory of this file
 * @param {*} filename The filename to fetch
 */
const rootDir = (filename) => {
    return path.resolve(__dirname, '..', '..', filename)
}


/**
 * Create a certificate by requesting a new one from the server and writing it to the root directory.
 */
const createCertificate = async () => {
    const response = await fetch(CERT_ENDPOINT, {
        method: 'POST'
    })

    if (response.status === 200) {
        const data = await response.json()

        const privateKey = data['private']
        const publicKey = data['public']
        const certificate = data['cert']

        fs.writeFileSync(rootDir('player.key'), privateKey)
        fs.writeFileSync(rootDir('player.pub'), publicKey)
        fs.writeFileSync(rootDir('player.crt'), certificate)
    } else {
        throw new Error('Got non-200 response from server')
    }
}


/**
 * Initialize the player certificate by checking if it exists.
 * If it does not, a new certificate will be fetched from the server
 * before starting.
 */
const initCertificate = async () => {
    key = rootDir('player.key')
    cert = rootDir('player.crt')

    if (!(fs.existsSync(key) && fs.existsSync(cert))) {
        await createCertificate()
    }
}


module.exports = {
    initCertificate: initCertificate
}
